app/mailers/user_notifications.rb

```diff
@@ -697,6 +685,51 @@ def send_notification_email(opts)
     build_email(user.email, email_opts)
   end
 
+  def self.participants(post, recipient_user, reveal_staged_email: false)
+    list = []
+
+    allowed_groups = post.topic.allowed_groups.order("user_count DESC")
+
+    allowed_groups.each do |g|
+      list.push("[#{g.name_full_preferred} (#{g.user_count})](#{g.full_url})")
+      break if list.size >= SiteSetting.max_participant_names
+    end
+
+    recent_posts_query = post.topic.posts
```

Comment:
Can we short-circuit here in case we already have enough participant names in order to avoid doing any unnecessary work?

Reply:
Yes, we could and I even thought about it, but decided against it for multiple reasons.
1. It's highly unlikely that there are 10 or more (the default setting) groups within a PM
2. We'd still need to query the number of allowed users without the recipient in order to calculate `others_count`

I think short-circuiting would add additional complexity and the performance gain is probably neglectable.